### PR TITLE
Clear WritableStreamSink queue on abort with compat flag

### PR DESCRIPTION
**Pull request by jasnell**
The motivation here is to address an issue with `WritableStream` instances wrapping a `WritableStreamSink` implementation not processing `abort()` calls proactively. Without the flag the `abort()` will be handled lazily, meaning that a consumer must be actively consuming the data written to the stream in order to trigger the abort and close the stream. This is problematic if the consumer has dropped or otherwise stopped reading from the writable's internal buffer ... the pending abort will never be processed and resolved. This means that pending writes and the pending abort itself will never be resolved in those cases. With the flag, the queue of pending writes is drained immediately when `abort()` is called and the stream is transitioned to an error state immediately. This should cause the underlying `WritableStreamSink` to be aborted and dropped immediately and the pending write and abort promises should all get settled.

internal ticket: EW-8607

### CODE REVIEW

**Comment by jasnell** [comment]
@anonrig ... just fyi on this PR. This is a good simple example of the introduction of a compat flag to preserve backwards compat behavior... just for illustration.

File: src/workerd/api/streams/common.h
Comment (kentonv) [inline src/workerd/api/streams/common.h:207]: Hmm, I usually associate the word "drain" with "gracefully complete all pending requests before shutting down", but here it seems to have the opposite meaning?

Maybe `FINISH_QUEUED_WRITES` and `CANCEL_QUEUED_WRITES` would be clearer names?

File: src/workerd/api/streams/internal.c++
Comment (kentonv) [inline src/workerd/api/streams/internal.c++:1172]: If there is currently a `write()` call in progress (that is, the `Promise` returned by a previous `write()` has not resolved yet), then this is UB, at least in the case of `EncodedAsyncOutputStream::abort()` which doesn't make any attempt to cancel that promise before deleting the underlying KJ stream.

One option is to change `WritableStreamSink`'s contract to say it is responsible for canceling writes on abort(). Another option would be to have `WritableStreamInternalController` itself be responsible for wrapping each write promise in a `kj::Canceler` which would be canceled here, before calling `abort()` on the sink. The latter seems better to me since it's handled in one place rather than every sink implementation.

Note that similarly, any `pumpTo()` needs to be canceled before aborting.

Comment (jasnell) [reply src/workerd/api/streams/internal.c++:1172]: Yep, was thinking along the same lines... I agree a kj::Canceler here would be a good approach.

Comment (jasnell) [reply src/workerd/api/streams/internal.c++]: Ok, updated but needs a careful review to make sure the new behavior is correct. It does not appear that we have any tests that verify the edge cases on this. I'll be looking as seeing if we can reasonably expand the test coverage here. But, in the meantime please take a look

**Review by anonrig — APPROVED** [review]

**Comment by jasnell** [comment]
Thank you for the review @anonrig ... going to wait for @kentonv to give another review before merging this.

File: src/workerd/api/streams/internal.c++
Comment (kentonv) [inline src/workerd/api/streams/internal.c++:1588]: Not new in this PR, but should we use `KJ_ASSERT_NONNULL(state.tryGet` here? `get` is UB if for whatever reason the state is a different type. IMO it should only ever be used if the type was checked literally right before.

File: src/workerd/api/streams/internal.c++
Comment (kentonv) [inline src/workerd/api/streams/internal.c++:2371]: Now that all the write requests are canceled first, is this needed?

Maybe we no longer need to pass `AbortOption` down to the `WritableStreamSink`?

File: src/workerd/api/system-streams.c++
Comment (kentonv) [inline src/workerd/api/system-streams.c++:303]: Interesting to note that for compressed streams, this implementation has always failed to flush the buffer before aborting, so the compat flag doesn't make a whole lot of difference (either way, previously-written data may be lost).

Makes me wonder if the compat flag is even needed.

**Comment by kentonv** [comment]
Sorry for delay, I had started reviewing this a few days ago and then got distracted away.

File: src/workerd/api/system-streams.c++
Comment (jasnell) [reply src/workerd/api/system-streams.c++:303]: Yeah that's a good point... hmmm

Comment (jasnell) [reply src/workerd/api/system-streams.c++:303]: I think the compat flag is still likely necessary. There's an observable difference in behavior around the rejected queued write promises here that someone could be currently working around ( at least in theory )...likely safest to have the flag

Comment (jasnell) [reply src/workerd/api/system-streams.c++:303]: but I think you're right about not needing the `AbortOption` enum class now.

**Review by kentonv — APPROVED** [review]

**Pull request merged**
